pulser-core/pulser/sequence/_seq_drawer.py

```diff
@@ -161,48 +163,47 @@ def _give_curves_from_samples(
         ]
 
 
-def gather_data(seq: pulser.sequence.Sequence, gather_output: bool) -> dict:
+def gather_data(
+    sampled_seq: SequenceSamples, shown_duration: Optional[int] = None
+) -> dict:
     """Collects the whole sequence data for plotting.
 
     Args:
-        seq: The input sequence of operations on a device.
-        gather_output: Whether to gather the modulated output curves.
+        sampled_seq: The samples of a sequence of operations on a device.
+        shown_duration: If present, is the total duration to be shown in
+            the X axis.
 
     Returns:
         The data to plot.
     """
     # The minimum time axis length is 100 ns
-    total_duration = max(
-        seq.get_duration(include_fall_time=gather_output), 100
-    )
+    total_duration = max(sampled_seq.max_duration, 100, shown_duration or 100)
     data: dict[str, Any] = {}
-    for ch, sch in seq._schedule.items():
-        # List of interpolation points
-        interp_pts: defaultdict[str, list[list[float]]] = defaultdict(list)
+    for ch, ch_samples in sampled_seq.channel_samples.items():
         target: dict[Union[str, tuple[int, int]], Any] = {}
         # Extracting the EOM Buffers
         eom_intervals = [
             EOMSegment(eom_interval[0], eom_interval[1])
-            for eom_interval in sch.get_eom_mode_intervals()
+            for eom_interval in ch_samples.get_eom_mode_intervals()
         ]
         nb_eom_intervals = len(eom_intervals)
         eom_start_buffers = [EOMSegment() for _ in range(nb_eom_intervals)]
         eom_end_buffers = [EOMSegment() for _ in range(nb_eom_intervals)]
         in_eom_mode = False
         eom_block_n = -1
         # Last eom interval is extended if eom mode not disabled at the end
-        if nb_eom_intervals > 0 and seq.get_duration() == eom_intervals[-1].tf:
+        if (
+            nb_eom_intervals > 0
+            and ch_samples.duration == eom_intervals[-1].tf
+        ):
             eom_intervals[-1].tf = total_duration
         # sampling the channel schedule
-        samples = sch.get_samples()
-        extended_samples = samples.extend_duration(total_duration)
-        for slot in sch:
-            if slot.ti == -1:
-                target["initial"] = slot.targets
-                continue
-            else:
+        extended_samples = ch_samples.extend_duration(total_duration)
+
+        for slot in ch_samples.slots:
+            if slot.ti != -1:
                 # If slot is not the first element in schedule
-                if sch.in_eom_mode(slot):
+                if ch_samples.in_eom_mode(slot):
                     # EOM mode starts
                     if not in_eom_mode:
                         in_eom_mode = True
```

Comment:
So indeed I would first try with `np.isclose(extended_samples.amp[slot.ti], 0)`. However, if the sample is the result of a modulated sample then it is not going to be close to 0 at `slot.ti`. I think it's best to check that `np.isclose(extended_samples.amp[slot.tf-1], 0)` or `(slot.ti + slot.tf)/2`.

Reply:
Neither one is working; the missing smooth_draw area in eom_end_buffers is from 940 to 1180, which was this delay _TimeSlot in the original schedule `_TimeSlot(type='delay', ti=940, tf=1180, targets={0, 1, 2, 3})`  (Still using EOM Mode Operation https://pulser.readthedocs.io/en/stable/tutorials/output_mod_eom.html as test case)
So the check on amp value should be correct, but it is not including the delay slot.

The output is restored as the original if I do this; if the current slot ti is different from the current eom_block end (tf), then it means there is a delay and we can add an end_buffer for this missing interval:

```python
                    if slot.ti != eom_intervals[eom_block_n].tf:
                        if extended_samples.amp[eom_intervals[eom_block_n].tf] == 0:
                            eom_end_buffers[eom_block_n] = EOMSegment(
                                eom_intervals[eom_block_n].tf, slot.ti
                            )
                    elif extended_samples.amp[slot.ti] == 0:
                        eom_end_buffers[eom_block_n] = EOMSegment(
                            slot.ti, slot.tf
                        )
```

I'm not sure if the elif part is still necessary and if this version is semantically identical to the original.

And this is the result:
![Figure_1](https://github.com/pasqal-io/Pulser/assets/1060425/1ce8d20f-a1eb-4052-a9ac-99f35d4bbd86)

Reply:
Good job finding this ! My question was related to using `draw_samples` on a modulated sample: imagine the user provides the hashed green curve to display (`draw_samples(sampler.sample(seq, modulation=True))`), how do you display the EOM mode and its buffer ? I think the first buffer will be shown, the main EOM part as well, but that you will have to check that your amplitude is not equal to 0 right after tf but after the fall time of the EOM (`2 * cast(BaseEOM, extended_samples._ch_objs[ch].eom_config).rise_time`, see in `ChannelSamples.modulate()`)
```
                    if slot.ti != eom_intervals[eom_block_n].tf:
                        if extended_samples.amp[eom_intervals[eom_block_n].tf + 2 * cast(BaseEOM, extended_samples._ch_objs[ch].eom_config).rise_time] == 0:
                            eom_end_buffers[eom_block_n] = EOMSegment(
                                eom_intervals[eom_block_n].tf, slot.ti
                            )
```

Reply:
You were right, the ending part was not show when drawing with `draw_samples(sampler.sample(seq, modulation=True))`: your solution works for the ending part.
The problem remains with the eom_start_buffers; the condition `extended_samples.det[slot.tf - 1] == ch_samples.eom_blocks[eom_block_n + 1].detuning_off` never becomes True, but the values are close:
-0.49731363802743167 instead of -0.5

Reply:
I think that using `np.isclose` and defining a `rtol` is definitely the way to go. Eventually I would make it dependent on the modulation_bandwidth (I think the higher the modulation bandwidth, the higher the error will be, so I guess you can define the relative tolerance as an empirical factor times the modulation bandwidth)

Reply:
Ok, I'll try making some experiment on it

Reply:
Doesn't the mod_bandwidth change only the rising time? I don't understand how it can affect the amplitude or the detuning